1.	Explain the key features of Python that make it a popular choice for programming?
Key Features:

•	Python is easy to understand and ease for learning.

•	Python has high cross platform compatibility than other programming languages.

•	Python has large variety of Standard Library.

•	Python has more readability and simplicity than other programming languages.

•	Python has been versatile and flexible than other programming languages.

•	Python has been Dynamically Typed.

•	Python has good Scalability and Performance than other programming languages.


2.	Describe the role of predefined keywords in Python and provide examples of how they are used in a programs ?
In Python, predefined key phrases are reserved phrases which have unique meanings and are used to define the shape and syntax of a program. They cannot be used as variable names or identifiers.

Examples of predefined keywords in Python include:

•	if, else, for, while (control flow)

•	def, class (function and class definitions)

•	import (module imports)

•	True, False, None (boolean and null values)

•	and, or, not (logical operators)


In [3]:
# Define a function using the `def` keyword
def greet(name):
    if name == "John":  # Use the `if` keyword for conditional statement
        print("Hello, John!")
    else:
        print("Hello, " + name + "!")

# Use the `for` keyword to iterate over a list
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)

# Use the `import` keyword to import a module
import math
print(math.pi)


apple
banana
cherry
3.141592653589793


3.	 Compare and contrast mutable and immutable objects in Python with examples

Mutable Objects:

Can be modified after creation.

Changes affect the original object.

Examples: lists, dictionaries, sets, user-defined objects .


In [1]:
my_dict = {"name": "John"}
my_dict["age"] = 30
print(my_dict)  # Output: {"name": "John", "age": 30}

{'name': 'John', 'age': 30}


Immutable Objects:

Cannot be modified after creation.

Changes create a new object.

Examples: strings, integers, floats, tuples.



In [2]:
my_string = "hello"
my_string += " world"
print(my_string)  # Output: "hello world" (new string created)

hello world


In [4]:
my_tuple = (1, 2, 3)
try:
  my_tuple[0] = 4
except TypeError:
  print("Cannot modify a tuple")

Cannot modify a tuple



Key differences:

Mutable objects can be modified in-place, while immutable objects create a new object when modified.

Mutable objects can be changed by multiple references, while immutable objects ensure that changes do not affect other references.

Best practices:

Use immutable objects when you need to ensure data integrity and avoid unintended changes.

Use mutable objects when you need to modify data in-place and improve performance.

Remember, in Python, everything is an object, and understanding the mutability of objects is crucial for writing efficient and effective code


Discuss the different types of operators in Python and provide examples of how they are used

Operators in Python

Python has several types of operators that allow you to perform various operations on variables and values. Here's a brief overview of each type, along with examples:

1.	 Arithmetic Operators
•	Perform mathematical operations
•	Examples:
            a = 5; b = 2; print(a + b) # Output: 7
            a = 5; b = 2; print(a ** b) # Output: 25

2.	 Comparison Operators
•	Compare values and return a boolean result
•	Examples:
            a = 5; b = 2; print(a > b) # Output: True
            a = 5; b = 2; print(a == b) # Output: False

3.	Logical Operators
•	Combine boolean values using AND, OR, and NOT
•	Examples: a = True; b = False; print(a and b) # Output: False
            a = True; b = False; print(a or b) # Output: True
            a = True; print(not a) # Output: False
4.	 Assignment Operators
•	Assign values to variables
•	Examples:
            a = 5 # Assign 5 to a
            a += 2 # Equivalent to a = a + 2
            a *= 3 # Equivalent to a = a * 3

5.	 Bitwise Operators
•	Perform binary operations on integers
•	Examples:
            a = 5; b = 3; print(a & b) # Output: 1
            a = 5; b = 3; print(a | b) # Output: 7
            a = 5; b = 3; print(a ^ b) # Output: 6

6.	 Membership Operators
•	Check if a value is present in a sequence (e.g., list, tuple, string)
•	Examples:
            my_list = [1, 2, 3]; print(2 in my_list) # Output: True
            my_string = "hello"; print("e" in my_string) # Output: True

7.	 Identity Operators
•	Check if two objects are the same
•	Examples:
            a = 5; b = 5; print(a is b) # Output: True
            a = [1, 2, 3]; b = [1, 2, 3]; print(a is b) # Output: False
	These are the main types of operators in Python. Understanding how to use them effectively is crucial for writing concise and efficient code.



5.	 Explain the concept of type casting in Python with examples

Type Casting in Python

Type casting, also known as type conversion, is the process of changing an object's data type from one type to another. In Python, you can perform type casting using built-in functions.

Here are some examples of type casting in Python:

int(x)
Converts a value to an integer:
x = 3.14
print(int(x))  # Output: 3

float(x)
Converts a value to a floating-point number:
x = 3
print(float(x))  # Output: 3.0

str(x)
Converts a value to a string:
x = 3
print(str(x))  # Output: "3"

How do conditional statements work in Python? Illustrate with examples

•	Conditional Statements in Python

Conditional statements in Python are used to execute different blocks of code based on certain conditions. Here's a brief overview of the three main types of conditional statements in Python, along with examples:

1.	If Statement

•	Executes a block of code if a condition is true:

    x = 5
    if x > 10:
       print("x is greater than 10")
    else:
       print("x is less than or equal to 10")
    # Output: x is less than or equal to 10

2.	 If-Else Statement
•	Executes one block of code if a condition is true, and another block if it's false:
            x = 5
            if x > 10:
                   print("x is greater than 10")
            else:
                  print("x is less than or equal to 10")
          # Output: x is less than or equal to 10

3.	 If-Elif-Else Statement
•	Executes one block of code if a condition is true, another block if a second condition is true, and a third block if both conditions are false:
            x = 5
            if x > 10:
                  print("x is greater than 10")
            elif x == 5:
                  print("x is equal to 5")
            else:
                  print("x is less than 5")
           # Output: x is equal to 5

4.	 Nested If Statements:
•	If statements can be nested inside each other:
            x = 5
            y = 10
            if x > 5:
                 if y > 10:
                       print("x is greater than 5 and y is greater than 10")
                 else:
                       print("x is greater than 5 but y is not greater than 10")
             else:
                   print("x is not greater than 5")
             # Output: x is not greater than 5


5.	Conditional Expressions:
•	A concise way to write simple if-else statements:
            x = 5
            result = "x is greater than 10" if x > 10 else "x is less than or equal to 10"
            print(result)
           # Output: x is less than or equal to 10

These are the basics of conditional statements in Python. By using if, elif, and else statements, you can control the flow of your program and make it more dynamic and responsive to different conditions.



Describe the different types of loops in Python and their use cases with examples.

Loops in Python

Loops in Python allow you to execute a block of code repeatedly for a specified number of iterations. Here are the three main types of loops in Python, along with their use cases and examples:

1.	For Loop
Used to iterate over a sequence (such as a list, tuple, or string) or other iterable objects:
            fruits = ['apple', 'banana', 'cherry']
            for fruit in fruits:
            print(fruit)
            # Output:
            # apple
            # banana
            # cherry
Use case: Iterate over a collection of items, such as processing a list of files or iterating over a database query result.

2.	 While Loop
Used to execute a block of code as long as a certain condition is true:
            i = 0
            while i < 5:
                  print(i)
                  i += 1
            # Output:
            # 0
            # 1
            # 2
            # 3
            # 4
Use case: Implement a loop that continues until a certain condition is met, such as reading input from a user until they enter a specific value.

3.	Nested Loops
Used to iterate over multiple sequences or perform complex iterations:
            for i in range(3):
                  for j in range(3):
                        print(f"i: {i}, j: {j}")

            # Output:
            # i: 0, j: 0
            # i: 0, j: 1
            # i: 0, j: 2
            # i: 1, j: 0
            # i: 1, j: 1
            # i: 1, j: 2
            # i: 2, j: 0
            # i: 2, j: 1
            # i: 2, j: 2
Use case: Iterate over multiple collections or perform complex iterations, such as generating a matrix or iterating over a list of lists.

These are the main types of loops in Python. By using for loops, while loops, and nested loops, you can perform a wide range of iterative tasks in your Python programs.



